### DISCUSSION TITLE
bug: vtsls error when using code actions

### DISCUSSION THREAD

**Discussion by redoxahmii** [LazyVim/LazyVim#3773] [Q&A]
### Did you check docs and existing issues?

- [X] I have read all the LazyVim docs
- [X] I have searched the existing issues of LazyVim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

0.10

### Operating system/version

Arch Linux

### Describe the bug

Prompts an error when using Code Actions in `.tsx` or `.jsx` files.
- This is the error : 
`  [Fzf-lua] "codeAction/resolve" failed with error -32603: Request codeAction/resolve failed with message: <sema 
 antic> TypeScript Server Error (5.4.5) 
 Debug Failure. Expected isNewFile for (only) new files. {"isNewFile":true,"hasScriptInfo":true} 
 Error: Debug Failure. Expected isNewFile for (only) new files. {"isNewFile":true,"hasScriptInfo":true} 
 at IpcIOSession.mapTextChangeToCodeEdit (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescr 
 ript/lib/tsserver.js:187338:13) 
 at /home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/tsserver.js:187330:46 
 at Array.map (<anonymous>) 
 at IpcIOSession.mapTextChangesToCodeEdits (/home/redox/Code/Experiments/next-nightwatch/node_modules/types 
 script/lib/tsserver.js:187330:25) 
 at IpcIOSession.getEditsForRefactor (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/ 
 /lib/tsserver.js:187207:21) 
 at getEditsForRefactor (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/tsserver. 
 .js:185408:43) 
 at /home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/tsserver.js:187569:69 
 at IpcIOSession.executeWithRequestId (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript 
 t/lib/tsserver.js:187561:14) 
 at IpcIOSession.executeCommand (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/t 
 tsserver.js:187569:29) 
 at IpcIOSession.onMessage (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/tsserv 
 ver.js:187611:51) 
 at process.<anonymous> (/home/redox/Code/Experiments/next-nightwatch/node_modules/typescript/lib/tsserver. 
 .js:189220:14) 
 at process.emit (node:events:519:28) 
 at emit (node:internal/child_process:951:14) 
 at process.processTicksAndRejections (node:internal/process/task_queues:83:21) `
 
 - visually how it looks:
 ![t_2024_06_22_07_36](https://github.com/LazyVim/LazyVim/assets/13983258/ab22bc30-2f92-4f99-8197-84c7d0badc79)

### Steps To Reproduce

1. Open a `.jsx` or `.tsx` file.
2. Run a Code Action

### Expected Behavior

Should not give this error.
Refactoring using the plugin or Code Actions does not really work that good when working with `tsx` or `jsx`.

### Repro

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  { "LazyVim/LazyVim", import = "lazyvim.plugins" },
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by arthur-mountain** [comment]
@Redoxahmii 
Could you please check your Node version first? You can do this by running `node -v`.

Additionally, please verify if the LSP server is running by checking for any vtsls errors in the `:LspLog`.

I have checked the vtsls repo that uses esbuild and supports up to Node 16. If your Node version is below 16, there may be syntax errors preventing the LSP server from running correctly.

**Reply by redoxahmii** [reply]
Node version is `v20.13.1` so that isn't the issue.
[This](https://0x0.st/XALq.txt) is my lsp.log to help.

**Reply by arthur-mountain** [reply]
@Redoxahmii 
If it's Node 20, there shouldn't be any syntax issues.

But let's check the LSP log. Regarding the vtsls part, it seems to be a TypeScript server error indicating that no project was found. You might want to see if there are any similar issues and solutions in the community.

Below are a few error stacks and messages I saw in the LSP log:
```
TypeScriptServerError: <semantic> TypeScript Server Error (5.4.5)\nNo Project.\nError: No Project.
```
```
serverMessage: 'No Project.',\n serverStack: 'Error: No Project.'
```
```
"TSServer exited. Code: null. Signal: SIGTERM"
```

**Reply by arthur-mountain** [reply]
Oh, but according to your error message from 'Describe the bug,' you are using tsserver instead of vtsls.

**Reply by dpetka2001** [reply]
`tsserver` is disabled when using the `Typescript` Extra, so maybe that's the reason? If you want to use explicitly `tsserver` then don't enable the Typescript Extra maybe?

**Reply by arthur-mountain** [reply]
Correct, when using typescript extra, then tsserver should be disabled. This is what confuses me.

I think we could give it a try

**Reply by dpetka2001** [reply]
`tsserver` is already disabled when using the Extra as can be seen [here](https://github.com/LazyVim/LazyVim/blob/8ea3f690e6017ba37022060a62f1d69fbf3aeaf7/lua/lazyvim/plugins/extras/lang/typescript.lua#L117-L120), that's why you get the error `"TSServer exited. Code: null. Signal: SIGTERM"`. Just disable the Typescript Extra and use your personal configuration where you're defining `tsserver` in `nvim-lspconfig` spec.

**Reply by redoxahmii** [reply]
`tsserver` is not even installed in mason this is what my `nvim-lspconfig` looks like.
```lua
{
    "neovim/nvim-lspconfig",
    event = "LazyFile",
    opts = {
      inlay_hints = { enabled = false },
    },
  }
  ```

**Reply by redoxahmii** [reply]
I think i know what triggers this error.
Whenever i select or even move to the option `Move to a new file` it throws this error even if i just use fuzzy find for that.
This happens only when using `fzf-lua` Extra but does not happen with `telescope`.

**Reply by dpetka2001** [reply]
I spoke based on how your [dotfiles look on Github](https://github.com/Redoxahmii/dotfiles/blob/8bc811b503a7391fb72d8a821e376c5320d30f4d/nvim/lua/plugins/utils/nvim-lspconfig.lua#L23-L49) (updated 2 days ago). I don't know if those also reflect your current dotfiles. It was just an assumption from my part. If I'm mistaken I do apologize.

**Reply by redoxahmii** [reply]
No problem atleast now we know what is triggering this error.
Can someone replicate this to check if they have the same issue when using with `fzf-lua` ?
Whenever the highlighted selector comes over `Move to new file` it triggers this error but only with `fzf-lua`.
EDIT: It does not trigger the very first time you do but the second time it will show this error.

**Reply by dpetka2001** [reply]
Unfortunately, I only do Python and Lua (started 6 months ago), so I'm not the best person to test this. Maybe you would like to ping the maintainer to see if he can shed more light into this? He's far more knowledgeable.

**Reply by redoxahmii** [reply]
`vtsls` does not seem to support `move to new file` atleast on my end it does not work and that somehow is triggering this error for `fzf-lua`

**Reply by redoxahmii** [reply]
@folke a little insight on this would be helpful.

**Reply by dpetka2001** [reply]
But with Telescope the code action does indeed move to a new file. Maybe `fzf-lua` does not support this code action? I just tested on a simple vanilla JS file with Telescope and it correctly completed the code action `move to new file`.

**Reply by redoxahmii** [reply]
Can confirm it works with `Telescope` the issue only occurs with `fzf-lua`.

**Reply by redoxahmii** [reply]
Code actions `Move to file` or `Move to new file`  do not really work properly in `jsx` or `tsx` at least as it sends the whole file there instead of the selected text but that seems to be lsp issue.

**Reply by dpetka2001** [reply]
@Redoxahmii So, I tested with `FzfLua lsp_code_actions previewer=false` and the problem doesn't occur. It seems there's something wrong with `fzf-lua` and the response from the LSP server (vtsls in this case) and there's some problem with the previewer. Can you confirm? If so, maybe you should open an issue at `fzf-lua` repo?

**Reply by dpetka2001** [reply]
My guess why it works with Telescope is because Telescope doesn't use code actions with preview. Something must be conflicting with the previewer. Not sure at all if this is a problem with `fzf-lua` itself or the response the LSP server sends back?

**Reply by redoxahmii** [reply]
@dpetka2001 can confirm that disabling the previewer the error does not occur.
i am also not sure if the issue is from `fzf-lua` side or just the LSP server.

**Reply by dpetka2001** [reply]
Not sure either, but it doesn't hurt to ask at `fzf-lua` repo. The maintainer will surely be able to guide you as to what the actual problem is. Whereas I'm just making guesses.

**Reply by redoxahmii** [reply]
Made an [issue](https://github.com/ibhagwan/fzf-lua/issues/1295) there so let's see what they say.

**Reply by ibhagwan** [reply]
This is an upstream issue, see my full reply https://github.com/ibhagwan/fzf-lua/issues/1295#issuecomment-2185268621, issue is similar to https://github.com/pmizio/typescript-tools.nvim/issues/231 (another `tsserver` wrapper), the `codeAction/resolve` request causes the next action (move to new file in our case) to fail, my guess is the server doesn't expect an LSP call in between the request for code actions and the call to `workspaceEdit`.

**Reply by redoxahmii** [reply]
Thank you for your input @ibhagwan.
@folke disabling the previewer as the default option should be considered.
I can make a PR for this but it's only a single [line](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/editor/fzf.lua#L180) that needs to be disabled and this will also make the behavior same to what Telescope has as default.

**Reply by redoxahmii** [reply]
One more thing is disabling this option makes the `Code Action` previewer look too big so that can also be reduced to use a smaller value on this [line](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/editor/fzf.lua#L132C1-L133C1)
Something like this looks good on my system but you can consider that how you want.
```lua
 height = math.floor(math.min(vim.o.lines * 0.3 - 16, #items + 2) + 0.5) + 16,
```
You can review these and if approved i can make the PR or you can do it  yourself.

**Reply by ibhagwan** [reply]
> Thank you for your input @ibhagwan. @folke disabling the previewer as the default option should be considered. I can make a PR for this but it's only a single [line](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/editor/fzf.lua#L180) that needs to be disabled and this will also make the behavior same to what Telescope has as default.

Idk about disabling the previewer due to one problmatic LSP, just my 2 cents.

**Reply by dpetka2001** [reply]
If it's only this one LSP, then it makes more sense to make your own mapping with no previewer for `js/ts` files specifically. Not all people face the prob you're facing, so disabling the previewer as a whole for just a specific code action from a LSP server sounds a bit much.

**Reply by redoxahmii** [reply]
Yes it can most definitely be centered down to use for only `js` and `ts` files. Disabling it completely was so it becomes consistent as `Telescope` is configured already.
Final decision is in the end to be made by the maintainer.

**Reply by ibhagwan** [reply]
It’s just for vtsls, the move file action works fine with tsserver directly with preview.

**Reply by dpetka2001** [reply]
It would be better if you made an issue at `vtsls` repo, so that people responsible for it could have a look at why this problem occurs and potentially fix it. In my personal opinion the previewer should just stay as it is and people who don't want it can make the appropriate changes in their personal configuration.

**Reply by dpetka2001** [reply]
We could also do in the fzf config file something like 
```lua
          code_actions = LazyVim.opts("nvim-lspconfig").servers.vtsls and {
            previewer = false,
          } or {
            previewer = vim.fn.executable("delta") == 1 and "codeaction_native" or nil,
          },
```
so that we explicitly only disable it for `vtsls` (since it doesn't work with it either way) and leave it enabled for the rest. Or the users just change it in their personal configuration. Not sure what the best approach should be.

**Reply by redoxahmii** [reply]
It is already mentioned on `vtsls` having issues with `codeAction/resolve` and is working in hacky way as you can see [here](https://github.com/yioneko/vtsls/issues/26).

I guess disabling it for only `vtsls` at the moment would be the better option as this is a known issue for a long time and i didn't see anything mentioning this being resolved any sooner.

**Reply by folke** [reply]
Would be great if someone could create a PR that hooks in the `ui_select` function of the fzf extra.
A simple check with `LazyVim.lsp.get_clients` for the current buffer. If `vtsls` is connected, then disable the preview.

**Reply by dpetka2001** [reply]
@Redoxahmii I submitted #3798 which hopefully solves this. It would be great if you could try it out.

**Reply by redoxahmii** [reply]
@dpetka2001 Still same issue on my end.
The previewer is not disabled for `vtsls` projects.

**Reply by dpetka2001** [reply]
Are you using LazyVim HEAD? Because it's not on LazyVim release yet.

**Reply by dpetka2001** [reply]
https://www.lazyvim.org/configuration/lazy.nvim

**Reply by redoxahmii** [reply]
@dpetka2001 yes it is working now.
